[Messenger] Adding support for record messages

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes 
| Fixed tickets | 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/10015

With inspiration from [this blog post](https://matthiasnoback.nl/2015/01/collecting-events-and-the-events-aware-command-bus/), I created a message recorder. This is needed when you handle messages in the same doctrine transaction. Ie, you have a `Command` and the `CommandHandler` dispatches `Events`. Those events should be handled in a new doctrine transaction. 

### Scenario

If the command handler creates a User with uuid X and dispatches `UserCreated` event. If they are handled in the same transaction then this code will find nothing: 

```php
$this->em->getRepository(User::class)->findOneBy(['uuid'=>X]);
```

---------

Example configuration:

```yaml

framework:
    messenger:
        default_bus: messenger.bus.command
        buses:
            messenger.bus.command:
                middleware:
                    # Make sure recorded events are handled by the event bus. You may also configure the message recorder with a second parameter. 
                    - app.messenger.middleware.record_messages
                    - doctrine_transaction_middleware: ['default']
        # -- more buses

services: 
    app.messenger.middleware.record_messages:
        parent: messenger.middleware.record_messages
        arguments: ['@messenger.bus.event']
```

Nyholm committed:
Adding support for record messages

Nyholm commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -296,6 +296,11 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
                 foreach (array_values($arguments ?? array()) as $key => $argument) {
+                    // Convert argument to references if needed.
+                    if (is_string($argument) && $argument[0] === '@') {
+                        $argument = new Reference(substr($argument, 1));
```
Ping @ogizanagi, This is the best way I found to support references. Im not sure if this is needed if you configure `framework.messenger` with XML.

Nyholm commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -296,6 +296,11 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
                 foreach (array_values($arguments ?? array()) as $key => $argument) {
+                    // Convert argument to references if needed.
+                    if (is_string($argument) && '@' === $argument[0]) {
+                        $argument = new Reference(substr($argument, 1));
```
> Ping @ogizanagi, This is the best way I found to support references. Im not sure if this is needed if you configure framework.messenger with XML.

Or should I go the long way of creating a Factory that uses `@container` as a dependency?

ogizanagi replied:
The idea of factories weren't to replace DI config when more suitable, only to solve most simple needs for configurable middlewares. Therefore, I don't think we should add services references support.
If really desired, we could indeed use dedicated service locators or perhaps let the user wiring it itself by creating the middleware service using DI config for now.

Nyholm replied:
Hm. I tried using a Factory that depends on `@service_container`. That does not work since the buses are private. 

I also tried injecting all services tagged with `message_bus` into the factory. That did not work because of circular dependencies.

>  Circular reference detected for service "debug.traced.messenger.bus.command", path: "debug.traced.messenger.bus.command -  
  > debug.traced.messenger.bus.command.inner -> messenger.bus.command.middleware.messenger.middleware.handles_recorded_mess  
  ages -> messenger.middleware.handles_recorded_messages.factory -> debug.traced.messenger.bus.command". 

Im saying that I do not know a better (working) solution that this. Im sorry.

ogizanagi replied:
Injecting the Symfony service container now belong to another era 😃 
We should register dedicated [service locators](https://symfony.com/blog/new-in-symfony-3-3-service-locators) instead. This would happen in the `MessengerPass`, using `ServiceLocatorTagPass::register()`.

But as I said, we could let the user create the concrete instance of this middleware in userland for now.

Nyholm replied:
Done. I've updated the configuration example

ogizanagi commented on src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php:
```diff
@@ -0,0 +1,51 @@
+        $this->messageRecorder->eraseMessages();
+
+        foreach ($recordedMessages as $recordedMessage) {
+            $this->messageBus->dispatch($recordedMessage);
```
What if a recorded message recorded another one?
Should we clear any recorded message before middleware execution, so it's obvious it's not supported & undesired by design, and no left over message is dispatched when reusing the bus to handle another message (am I clear? 😅).

Also should an exception while dispatching a recorded message really interrupt everything?

Nyholm replied:
Interesting. 

I think I agree with all your questions. I'll try to make some changes to address them.

Killium commented on src/Symfony/Component/Messenger/MessageRecorderInterface.php:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @return object[]
+     */
+    public function recordedMessages(): array;
```
Why not naming this function `fetch()` instead of `recordedMessages()`, as you already name below the method `record()` (without *messages* at the end of the name) ?

Same thing for `eraseMessages()`.

Nyholm replied:
I took the names from SimpleBus. But you are correct, your naming seams easier

Nyholm commented on src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php:
```diff
@@ -0,0 +1,66 @@
+                try {
+                    $this->messageBus->dispatch($recordedMessage);
+                } catch (\Throwable $exception) {
+                    $exceptions[] = $exception;
```
Note that if one handler to message A throws an exception, then other handlers may not run (depending on the order).

This will just make sure that handlers to message B will run even though a handler to A throws an exception.

kbond replied:
This slightly ties into what I proposed in #27215. I think all handlers should be run even if some fail.

ogizanagi commented on src/Symfony/Component/Messenger/MessageRecorder.php:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function erase(): void
```
`reset`? Would be a legit use-case for the `ResettableInterface` too I guess.

ogizanagi commented on src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        $exceptions = array();
+        while(!empty($recordedMessages = $this->messageRecorder->fetch())) {
```
`if` rather than `while`?

ogizanagi commented on src/Symfony/Component/Messenger/Exception/MessageRecorderHandlerException.php:
```diff
@@ -0,0 +1,40 @@
+    public static function create(array $exceptions): self
+    {
+        $message = sprintf(
+            "One or more handlers for reordered messages threw an exception. Their messages were: \n\n%s",
```
`s/reordered/recorded`

ogizanagi commented on src/Symfony/Component/Messenger/Exception/MessageRecorderHandlerException.php:
```diff
@@ -0,0 +1,40 @@
+    {
+        return $this->exceptions;
+    }
+}
```
Missing new line

Nyholm replied:
Fixed

ogizanagi commented on src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        if (!empty($exceptions)) {
+            throw MessageRecorderHandlerException::create($exceptions);
```
We don't use named construct for exceptions in core ^^

Wild thoughts:
Should it really throw? Make it configurable? Should the `MessageRecorder` store the exceptions itself?

Nyholm replied:
Hm.. I think we should throw. 
But making it configurable sounds like a reasonable future update.

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -34,6 +34,11 @@
             <argument type="service" id="validator" />
         </service>
 
+        <service id="messenger.middleware.handles_recorded_messages" class="Symfony\Component\Messenger\Middleware\HandlesRecordedMessagesMiddleware" abstract="true">
```
Looking at the names of other middlewares, this one maybe should be called `messenger.middleware.record_messages`

Killium commented on src/Symfony/Component/Messenger/MessageRecorderInterface.php:
```diff
@@ -0,0 +1,38 @@
+    public function fetch(): array;
+
+    /**
+     * Erase messages that were recorded since the last call to eraseMessages().
```
As your function is called `reset()`, I would have say:

> Reset messages that were recorded since the last call to eraseMessages().

ogizanagi replied:
> Reset messages that were recorded since the last call to **reset()**.

😄

sroze replied:
You can actually extend the `ResettableInterface` from `symfony/contracts` now.

Killium commented on src/Symfony/Component/Messenger/MessageRecorderInterface.php:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @param object $message
+     */
+    public function record($message);
```
Add `: void` as function signature (because of interface).

sroze replied:
I don't see any value in adding `: void` tbh.

sroze commented on src/Symfony/Component/Messenger/Exception/MessageRecorderHandlerException.php:
```diff
@@ -0,0 +1,41 @@
+namespace Symfony\Component\Messenger\Exception;
+
+/**
+ * When handling recorded messaged one or more handlers caused an exception.
```
`s/one/on`

sroze commented on src/Symfony/Component/Messenger/Exception/MessageRecorderHandlerException.php:
```diff
@@ -0,0 +1,41 @@
+    public function __construct(array $exceptions)
+    {
+        $message = sprintf(
+            "One or more handlers for recorded messages threw an exception. Their messages were: \n\n%s",
```
If there is only one exception, let's not confuse the user and display it without wrapping it to "one or more".

sroze commented on src/Symfony/Component/Messenger/Exception/MessageRecorderHandlerException.php:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class MessageRecorderHandlerException extends \RuntimeException implements ExceptionInterface
```
`MessageHandlingException`? This has nothing to do with the "recorder" AFAIK

sroze commented on src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php:
```diff
@@ -0,0 +1,66 @@
+
+        $exceptions = array();
+        while (!empty($recordedMessages = $this->messageRecorder->fetch())) {
+            $this->messageRecorder->erase();
```
If the use-case is always `fetch` and `erase`... I'm pretty sure we can find better than `fetch` + `erase/reset`.

Nyholm replied:
Im not sure what you mean here.

sroze commented on src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php:
```diff
@@ -0,0 +1,66 @@
+    public function handle($message, callable $next)
+    {
+        // Make sure the recorder is empty before we begin
+        $this->messageRecorder->erase();
```
Why would you? 🤔

Nyholm replied:
To avoid any side effects. See https://github.com/symfony/symfony/pull/27844#discussion_r200204703

sroze reviewed (dismissed):
In this pull-request, there is nothing that actually records the messages, right? 🤔

As far as I understand, the problem you are trying to solve is to wrap messages into some "transaction", so that dispatched message will be handled after the "current" one, right? If I understand it correctly then I believe we should work on the naming to clarify this and we can use the term "transactions", it isn't necessarily database-only.

Nyholm commented:
> In this pull-request, there is nothing that actually records the messages, right? 🤔

True. "record" is a word from SimpleBus. 

> As far as I understand, the problem you are trying to solve is to wrap messages into some "transaction", so that dispatched message will be handled after the "current" one, right? If I understand it correctly then I believe we should work on the naming to clarify this and we can use the term "transactions", it isn't necessarily database-only.

Correct. I want to wrap my *commands* in one transaction and handle my *events* later. So when used with doctrine transaction middleware, if my event fails, it should not roll back the command.

gubler commented:
I implemented this pull request in an application I'm working on after it was pointed out to me by @ogizanagi in Slack and @sroze asked me to write up a use case.

I am working on App with a Command Bus with doctrine transactions and had an Event Bus. When injecting the Event Bus directly to the Command Handler, any error in the doctrine transaction caused a rollback as it was supposed to, but the Events were still dispatched to the Event Bus.

I copied the contents of this pull request into my app's namespace - just changing the namespaces for everything and it worked flawlessly. I did have to add the `reset` method to the `MessageRecorderInterface` because I am on Symfony 4.1 and not using `symfony/contracts`.

Now I inject the message recorder and if my transaction fails, the event is not dispatched. Here is an example Command Handler:

```php
namespace App\Domain\CommandHandler\Facility;

use App\Domain\Command\Facility\NewFacilityCommand;
use App\Domain\Event\Facility\FacilityCreatedEvent;
use App\Lib\Messenger\MessageRecorder\MessageRecorderInterface;
use App\Repository\FacilityRepository;

/**
 * Class NewFacilityHandler.
 */
class NewFacilityHandler
{
    /** @var FacilityRepository */
    protected $repository;
    /** @var MessageRecorderInterface */
    protected $recorder;

    /**
     * @param FacilityRepository       $repository
     * @param MessageRecorderInterface $recorder
     */
    public function __construct(FacilityRepository $repository, MessageRecorderInterface $recorder)
    {
        $this->repository = $repository;
        $this->recorder = $recorder;
    }

    /**
     * @param NewFacilityCommand $command
     */
    public function __invoke(NewFacilityCommand $command)
    {
        $facility = $this->repository->persist($command);

        $this->recorder->record(
            new FacilityCreatedEvent(
                $facility,
                $command->generator
            )
        );
    }
}
```

I kept all the names the same so when this is (hopefully) merged and added to a tagged release, I can update all of the `use` statements and my `services.yaml` file and everything should continue to work the same.

I did have one problem until I read the documentation PR that mentioned that the message recorder middleware needed to be declared _before_ the transaction middleware, but once I fixed that, it worked fine.

Here is my `messages.yaml` if it helps:

```yaml
framework:
    messenger:
        default_bus: messenger.bus.commands

        buses:
            messenger.bus.commands:
                middleware:
                    - validation
                    - record_messages
                    - doctrine_transaction_middleware
                    - 'App\Domain\BusMiddleware\SecurityMiddleware'
            messenger.bus.events:
                middleware:
                    - validation
                    - allow_no_handler
                    - doctrine_transaction_middleware
                    - 'App\Domain\BusMiddleware\SecurityMiddleware'
                    - 'App\Domain\EventBusMiddleware\FlashMessageMiddleware'
            messenger.bus.queries:
                middleware:
                    - validation
                    - doctrine_transaction_middleware
                    - 'App\Domain\BusMiddleware\SecurityMiddleware'
```

and the relevant parts of my `services.yaml`. I had to add the `method` argument to the tags for `message.recorder` - I assume this is bacause I am using Symfony 4.1 and not using `symfony/contracts`.

```yaml
 # doctrine middleware for command bus
    doctrine.orm.messenger.middleware_factory.transaction:
        class: Symfony\Bridge\Doctrine\Messenger\DoctrineTransactionMiddlewareFactory
        arguments: ['@doctrine']

    messenger.middleware.doctrine_transaction_middleware:
        class: Symfony\Bridge\Doctrine\Messenger\DoctrineTransactionMiddleware
        factory: ['@doctrine.orm.messenger.middleware_factory.transaction', 'createMiddleware']
        abstract: true
        arguments: ['default']

    messenger.middleware.record_messages:
        class: App\Domain\BusMiddleware\HandlesRecordedMessagesMiddleware

    messenger.recorder:
        class: App\Lib\Messenger\MessageRecorder\MessageRecorder
        tags:
            - name: 'kernel.reset'
              method: 'reset'

    App\Lib\Messenger\MessageRecorder\MessageRecorderInterface:
        alias: "messenger.recorder"
```

arnolanglade commented on src/Symfony/Component/Messenger/MessageRecorderInterface.php:
```diff
@@ -0,0 +1,35 @@
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Matthias Noback <matthiasnoback@gmail.com>
+ */
+interface MessageRecorderInterface extends ResetInterface
```
SimpleBus split this interface into 2 small ones. The first one `ContainsRecordedMessages` is implement by our models because you need to tell them to give us the events raise during the transaction. The second one `RecordsMessages` is used to record event at this end of the transaction. For instance, I use it in my repository when my aggregate is flushed. I guess this two distinct responsibilities: I don't want that my model have a public method `record`.

arnolanglade replied:
BTW, I prefer the naming of simple bus (declarative vs imperative programming). It is better to describer what happens instead how it happens :)

Nyholm replied:
I thought about slitting them. But I was not sure. Thank you for commenting!

arnolanglade commented on src/Symfony/Component/Messenger/MessageRecorderInterface.php:
```diff
@@ -0,0 +1,35 @@
+     *
+     * @return object[]
+     */
+    public function fetch(): array;
```
This name seems weird to me. Like I said previously my model will implement this interface. For example, a product model. My product will have Product::fetch() and Product::reset(). In term of business that does not have any sense. I think we should use `eraseMessage()` and `recordedMessage()` it more explicit. I think it is a good compromise between technical and business stuff.

arnolanglade commented on src/Symfony/Component/Messenger/Exception/MessageHandlingException.php:
```diff
@@ -0,0 +1,41 @@
+        parent::__construct($message);
+    }
+
+    public function getExceptions(): array
```
`toArray` would be better.

Nyholm replied:
Hm, Im not sure

arnolanglade commented on src/Symfony/Component/Messenger/MessageRecorder.php:
```diff
@@ -0,0 +1,45 @@
+ */
+class MessageRecorder implements MessageRecorderInterface
+{
+    private $messages = array();
```
Don't the Symfony standard recommend to use [] ?

Nyholm replied:
No, the other way around. Symfony is always using the long array syntax.

arnolanglade commented on src/Symfony/Component/Messenger/Middleware/HandlesRecordedMessagesMiddleware.php:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        if (!empty($exceptions)) {
+            if (1 === count($exceptions)) {
```
Why do you do that? I think it will be better to throw a single type of exception. It will be easier to handle it.

Nyholm replied:
The `MessageHandlingException` is only for grouped exceptions. If the group only have one element, there is no need for a group. 

I think an earlier review suggested it to be this way.

arnolanglade commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -34,6 +34,11 @@
         </service>
 
+        <service id="messenger.middleware.record_messages" class="Symfony\Component\Messenger\Middleware\HandlesRecordedMessagesMiddleware" abstract="true">
+            <argument type="service" id="message_bus" />
```
`message_bus`? Will you dispatch the events into the same bus than the command?

Nyholm replied:
This will be overwritten by your configuration. It is just abstract.

Nyholm commented:
Thank you for the reviews. I've made some updates.

Nyholm commented:
Of course. I'll add some tests. 
Also note the PR to the docs: https://github.com/symfony/symfony-docs/pull/10015

Nyholm commented:
@chalasr, I've added some tests now. It does not really make any sense adding tests to the framework bundle (at least not like that you referred to). Because this PR only registers two new services in the framework bundle. 

I added unit tests for my code though.

arnolanglade commented on src/Symfony/Component/Messenger/MessageRecorder.php:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function reset()
```
Why do you keep this method? You should remove it. Like I said, `reset`does not mean anything for a domain object.

Nyholm replied:
I agree. But I need this service to implement `ResettableInterface`. You do not need to implement this method in your domain objects. I removed it from the interface.  =)

arnolanglade replied:
ok :+1:

arnolanglade commented on src/Symfony/Component/Messenger/MessageRecorder.php:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function getRecordedMessages(): array
```
I am not fan of `getter`, it does not follow the `tell, don't ask` principle. IMOH `recordedMessages` is enough/better :)

Nyholm replied:
As far as I know, this is the "symfony way". Im not sure if I can apply "`tell, don't ask` principle". Even though we may like it more at the moment.

arnolanglade replied:
ok I understand :(

arnolanglade commented on src/Symfony/Component/Messenger/RecordedMessageContainerInterface.php:
```diff
@@ -0,0 +1,31 @@
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Matthias Noback <matthiasnoback@gmail.com>
+ */
+interface RecordedMessageContainerInterface
```
I am not sure about Container, it is just a simple collection. `RecordedMessagesInterface` should be enough IMHO.

Nyholm replied:
Hm.. I'm hesitant. Then we have two interfaces with super similar namses: `RecordedMessagesInterface` and `MessageRecorderInterface`. 
I wanted to use `RecordedMessageContainerInterface` so it is more explicit.

Nyholm replied:
Perfect

arnolanglade commented on src/Symfony/Component/Messenger/RecordedMessageContainerInterface.php:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @return object[]
+     */
+    public function getRecordedMessages(): array;
```
same feedback about `getter`.

chalasr commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:
```diff
@@ -34,6 +34,11 @@
         </service>
 
+        <service id="messenger.middleware.record_messages" class="Symfony\Component\Messenger\Middleware\HandleRecordedMessageMiddleware" abstract="true">
+            <argument type="service" id="message_bus" />
```
Probably not a big deal, but I would use `<argument type="service" /> <!-- Message bus -->` to make it fully abstract

Nyholm replied:
Yes, But if you only have one bus configured, do you really want to explicit define it again in the config? Or maybe that is a good idea so the user understands what is happening.

arnolanglade commented on src/Symfony/Component/Messenger/MessageRecorder.php:
```diff
@@ -0,0 +1,55 @@
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Matthias Noback <matthiasnoback@gmail.com>
+ */
+class MessageRecorder implements MessageRecorderInterface, RecordedMessageContainerInterface, ResettableInterface
```
What do you think about creating a trait like `SimpleBus`. This is useful, especially for domain model.

Nyholm replied:
I agree. We should also create a doctrine event subscriber that implements `RecordedMessageContainerInterface`. But that should be a separate PR. Do you want to create that those two?

arnolanglade replied:
Yes we can create create it in another PR. I can do it! I don't get you about creating a doctrine subscriber.

Nyholm replied:
Awesome. Please ping me in that PR and I'll be happy to review. 

About doctrine subscriber. We need (just like simplebus) something that takes messages from your domain objects and add them to a message bus.

Nyholm commented:
Status: Needs Review

Koc commented:
@Nyholm consider another approach, when aggregate root emits events and they are dispatched after transaction commited. See https://beberlei.de/2013/07/24/doctrine_and_domainevents.html for more details.

With current approach we need emit events manually from command handlers.

Nyholm commented:
That is a follow up PR :)
It is just a new listener and a messageContainer :)

> On 15 Sep 2018, at 04:08, Konstantin Myakshin <notifications@github.com> wrote:
> 
> @Nyholm consider another approach, when aggregate root emits events and they are dispatched after transaction commited. See https://beberlei.de/2013/07/24/doctrine_and_domainevents.html for more details.
> 
> With current approach we need emit events manually from command handlers.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Nyholm commented:
I've squashed and rebased. I'm ready for a final review.

sroze commented on src/Symfony/Component/Messenger/Middleware/HandleRecordedMessageMiddleware.php:
```diff
@@ -0,0 +1,71 @@
+    public function handle($message, callable $next)
+    {
+        // Make sure the recorder is empty before we begin
+        $this->messageRecorder->resetRecordedMessages();
```
Let's imagine the following example:

1. `Controller` `dispatch`es a `FirstMessage` message.
2. `FirstMessageHandler` gets called and it does the following:
    ```php
    $recorder->record(new SecondMessage);
    $bus->dispatch(new SaveToSecondaryDatabase);
    $recorder->record(new ThirdMessage);
    ```

The `dispatch` will call the middleware and the just recorded message will be erased, right?

Nyholm replied:
This is correct IF the current bus and `$bus` is using the same `MessageRecorder`. 

I think I should remove this line. Are you okey with that @ogizanagi?  It was added by [your suggestion](https://github.com/symfony/symfony/pull/27844#discussion_r200204703).

ogizanagi replied:
It might be fine removing it, but the case described above is a more complex topic then.
For instance, if there is any issue with `FirstMessageHandler` after `SaveToSecondaryDatabase`, `SecondMessage` 
 (and possibly those recorded by `SaveToSecondaryDatabase`) would still have been executed while I would expect it does only if `FirstMessageHandler` is fully successful.
Actually, to me recorded messages should be treated if and only if the highest `dispatch` call is successful. Possibly by using an internal flag inside the middleware? (Thus, we could even keep the reset for safety, only when the flag is off).

ogizanagi replied:
Perhaps the `LockingMiddleware` of Tactician could be a source of inspiration: 
- http://tactician.thephpleague.com/plugins/locking-middleware/
- https://github.com/thephpleague/tactician/blob/master/src/Plugins/LockingMiddleware.php

sroze replied:
Which is equivalent to [the `FinishesHandlingMessageBeforeHandlingNext` middleware](https://github.com/SimpleBus/SimpleBus/blob/master/Component/MessageBus/src/Bus/Middleware/FinishesHandlingMessageBeforeHandlingNext.php) of SimpleBus. (I prefer SimpleBus' naming in here, it's more explicit)

sroze commented on src/Symfony/Component/Messenger/Middleware/HandleRecordedMessageMiddleware.php:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        $exceptions = array();
+        while (!empty($recordedMessages = $this->messageRecorder->getRecordedMessages())) {
```
What's the reasoning of `get` + `reset` instead of `pop`?

Nyholm replied:
I want to make sure nobody runs `$this->messageRecorder->resetRecordedMessages();` while I'm handling the recorded messages.

Nyholm commented:
I rebased and everything is green.

weaverryan commented:
@Nyholm is the documentation PR more-or-less up-to-date with the most recent changes? I'm want to "wrap my head" around this PR - I was starting by reading your docs :)

Nyholm commented:
Yes. It’s up to date.

javiereguiluz commented:
@Nyholm is *"record"* a common/required word in this context? Personally I would always say: *"save messages"* or *"store messages"* to refer to something like this but never *"record messages"*.

Nyholm commented:
That is a fair point. "Record" is used from SimpleBus. As far as I know it is made up by them. Using "record" will make it easier for SimpleBus users to migrate (if they would like to).

Im all for using another word but it has to be more clear than using "record". I think that using "Store" or "Save" would cause for confusion since the component support adding things to a queue or "queue like". 

Since I'm not "saving" any messages anywhere (just having them in a temporary memory queue) I don't think it is a good word.

sroze commented:
Then, what about this?
```php
$this->bus->dispatch(Envelope::wrap($message)->with(new ToBeDispatchedAfterCurrentMessageHandling())
```

All we need is a middleware that is doing the "locking" (i.e. tactitian term) or "FinishesHandlingMessageBeforeHandlingNext" (SimpleBus) only if the following envelope item is present.

Nyholm commented:
With an EventRecorder you can collect events from Doctrine (just as @arnolanglade is doing). 

```php
class User {
  use EventRecorderTrait; // To be created in a following PR

  public function setUsername($username) 
  { 
    $this->username = $username;
    $this->record(new UsernameWasUpdated());
  }
  // ...
}
```
Using the minor modification in my entity as shown above and a `DoctrineEventRecorder` (Doctrine subscriber) I can make sure that domain events like these eventually ends up on the bus. 

--------

You will also have the flexibility to decide which events that should be recorded and which should be handled now. 

```php
public function handle(FooCommand $command)
{
    $this->recorder->record(new Bar()); // Handle later
    $this->eventBus->dispatch(new Biz()); // Handle now
}
```

This is not possible with a middleware like `FinishesHandlingMessageBeforeHandlingNext`.

-------------

I know "`EventRecorder`" and "record" is a new concept to this component. But it is not a new concept for message busses. SimpleBus has successfully been using this for years.

sroze commented:
> You will also have the flexibility to decide which events that should be recorded and which should be handled now.

That would exactly be the point of this Envelope item [in this comment](https://github.com/symfony/symfony/pull/27844#issuecomment-423748836), right?

> Using the minor modification in my entity as shown above and a DoctrineEventRecorder (Doctrine subscriber) I can make sure that domain events like these eventually ends up on the bus.

But you could do the same with the other option which does not introduce this extra notion of "recorder", right?

Nyholm commented:
No, you can’t. (As far as I know). How and when would you dispatch new messages to the bus? 
Sure, you can create a middleware called “DispatchMessagesCollectedByDoctrineSubscriber”. That would possibly achieve the same thing but will less abstraction and flexibility. 

I don’t know why we are hesitant of introducing this concept to the MessengerComponent. The argument cannot be DX and/or to keep the component simple, because the alternative solutions proposed are neighter simpler or more intuitive. 

Btw, I like that we are challenging this, and I don’t think we should have redundant ways of doing the same thing.

Koc commented:
How it possible to use this `MessageRecorder` for event sourcing and store all triggered events in db in **same** transaction? But queue them to the bus after transaction is complete.

Nyholm commented:
This is how you keep them in different transaction:

```yaml
# config/packages/messenger.yaml
framework:
    messenger:
        default_bus: messenger.bus.command
        buses:
            messenger.bus.command:
                middleware:
                    - messenger.middleware.validation
                    - messenger.middleware.handles_recorded_messages
                      # Doctrine transaction must be after handles_recorded_messages middleware
                    - doctrine_transaction_middleware: ['default']
            messenger.bus.event:
                middleware:
                    - messenger.middleware.allow_no_handler
                    - messenger.middleware.validation
```

If you change the order fo the middlewares, you can get them in the same middleware: 

```yaml
            messenger.bus.command:
                middleware:
                    - messenger.middleware.validation
                    - doctrine_transaction_middleware: ['default']
                    - messenger.middleware.handles_recorded_messages
````

----

If you are talking about events from a CommandHandler you would obviously not use this event recorder but the event bus directly. But if you are referring to the events from your entities, then the solution above it for you. 

If you want to use both, just define more services for the same classes.

Koc commented:
> If you are talking about events from a CommandHandler you would obviously not use this event recorder but the event bus directly.

No, I cann't. Command handing wrapped in transaction and we shouldn't send any events until it commited.

1. emit events in aggregate root
2. save aggregate root and emited events in db in same transaction
3. after transaction commited - send emited events also to bus.

> If you want to use both, just define more services for the same classes.
Not sure that I understand what services should be defined twice. Can you elaborate?

fabpot commented:
Closing in favor of #28849
